internal/command/clusters/uninstall.go

```diff
@@ -0,0 +1,349 @@
+		Short: "Check that a cluster is ready to have Jetstack Software uninstalled",
+		Long: `Runs the following checks:
+* Checks secrets containing certificates are safe from garbage collection
+* Checks for any upcoming renewals
+* Checks for certificates currently being issued
+* Checks for certificates that will expire soon
+* Checks for unready certificates
+* Checks for failing issuances
+`,
+		Args: cobra.MatchAll(cobra.ExactArgs(0)),
+		Run: run(func(ctx context.Context, args []string) error {
+			kubeCfg, err := kubernetes.NewConfig(kubeConfigPath)
+			if err != nil {
+				return err
+			}
+
+			clientset, err := buildClients(kubeCfg)
+			if err != nil {
+				return fmt.Errorf("error building required clients: %w", err)
+			}
+
+			realClock := clock.RealClock{}
+			notifications, err := findIssues(ctx, clientset, realClock)
+			if err != nil {
+				return fmt.Errorf("error investigating cluster state: %w", err)
+			}
+
+			// print out any suggested next steps
+			if len(notifications) > 0 {
+				fmt.Fprintf(os.Stdout, "\nResults:\n")
+				for _, n := range notifications {
+					fmt.Fprintf(os.Stdout, "%s\n", n.header)
+					for _, ri := range n.resourceInfos {
+						fmt.Fprintf(os.Stdout, "	* %s\n", ri)
+					}
+				}
+			} else {
+				fmt.Fprintf(os.Stdout, "\nNothing to do before uninstalling\n")
+			}
+			return nil
+		}),
+	}
+
+	return cmd
+}
+
+func findIssues(ctx context.Context, clientset allClients, clock clock.Clock) ([]notification, error) {
+	notifications := []notification{}
+	nowTime := clock.Now()
+
+	// Check all cluster Secrets for potential issues
+	var secretsList corev1.SecretList
+	err := clientset.secrets.List(ctx, &clients.GenericRequestOptions{}, &secretsList)
+	if err != nil {
+		return nil, fmt.Errorf("error listing cluster secrets: %w", err)
+	}
+
+	fmt.Fprintf(os.Stdout, "Running checks against cluster Secrets:\n")
+	fmt.Fprintf(os.Stdout, "	* Checking that issued certificates are safe from garbage collection...\n")
+	ownerRefsNotification := notification{}
+	for i := range secretsList.Items {
+		secret := &secretsList.Items[i]
+
+		hasCertificateOwnerRef := false
+		for _, ownerRef := range secret.OwnerReferences {
+			if ownerRef.Kind == cmapi.CertificateKind {
+				hasCertificateOwnerRef = true
+				break
+			}
+		}
+
+		if hasCertificateOwnerRef {
+			ownerRefsNotification.resourceInfos = append(ownerRefsNotification.resourceInfos, fmt.Sprintf(hasOwnerRefInfoTemplate, secret.Namespace, secret.Name))
+		}
+	}
+
+	if len(ownerRefsNotification.resourceInfos) > 0 {
+		ownerRefsNotification.header = hasOwnerRefHeader
+		notifications = append(notifications, ownerRefsNotification)
+	}
+	var certificates cmapi.CertificateList
+	err = clientset.certificates.List(
+		ctx,
+		&clients.GenericRequestOptions{},
+		&certificates,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error listing certificates: %s", err)
+	}
+
+	// Check all cluster Certificates for potential issues
+	unreadyNotification := notification{}
```

Comment:
Instead of creating these notifications here already, can we only create them when adding to the notifications list? And just use []string here?

Reply:
Why?

Reply:
Its somewhat confusing that we are creating these notification objects and only set the header when they are being added to the notifications array.

Reply:
fixed in https://github.com/jetstack/jsctl/pull/71/commits/6478dbd89f3b4e8a18eca5421e6f76dfef42e3e8